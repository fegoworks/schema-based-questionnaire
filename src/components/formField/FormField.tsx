import { FunctionComponent } from "react";
import { Checkbox, RadioGroup, SelectField, TextField } from "..";
import { FormikTouched } from "formik";

export type Opt = { value: string | number; label: string };

export interface FieldProps {
  error?: any;
  name: string;
  onBlur?: (e: any) => void;
  onChange?: (e: any) => void;
  touched?: boolean | FormikTouched<any> | FormikTouched<any>[] | undefined;
  type?: string;
  value?: string;
  placeholder?: string;
  options?: Opt[];
  checked?: boolean;
  disabled: boolean;
}

const FormField: FunctionComponent<FieldProps> = ({ type, ...props }) => {
  switch (type) {
    case "text":
    case "number":
    case "date":
      return <TextField type={type} {...props} />;

    case "select":
      return <SelectField {...props} />;
    case "radio-group":
      return <RadioGroup {...props} />;

    case "checkbox":
      return <Checkbox {...props} />;
    default:
      return <div></div>;
  }
};

export default FormField;
