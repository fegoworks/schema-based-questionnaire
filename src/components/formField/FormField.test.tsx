import { render } from "@testing-library/react";
import FormField from "./FormField";

const setup = () => {
  const defaultProps = {
    name: "name",
    disabled: false,
  };

  return {
    defaultProps,
  };
};

describe("Component - FormField", () => {
  it("renders FormField component", () => {
    const { defaultProps } = setup();
    const { container } = render(<FormField type="text" {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("renders textField component when the type is text, number or date", () => {
    const { defaultProps } = setup();
    [
      { type: "text", name: "FirstName" },
      { type: "number", name: "PhoneNumber" },
      { type: "date", name: "Date" },
    ].forEach((input) => {
      const { container } = render(
        <FormField type={input.type} {...defaultProps} name={input.name} />
      );
      const textFields = container.getElementsByClassName("TextField");
      expect(textFields).toHaveLength(1);
    });
  });

  it("renders selectField component when the type is select", () => {
    const { defaultProps } = setup();
    const { container } = render(
      <FormField type="select" {...defaultProps} name="Companies" />
    );
    const selectFields = container.getElementsByClassName("SelectField");
    expect(selectFields).toHaveLength(1);
  });

  it("renders checkbox component when the type is checkbox", () => {
    const { defaultProps } = setup();
    const { container } = render(
      <FormField type="checkbox" {...defaultProps} name="Companies" />
    );
    const checkboxField = container.getElementsByClassName("Checkbox");
    expect(checkboxField).toHaveLength(1);
  });

  it("renders radioGroup component when the type is RadioGroup", () => {
    const { defaultProps } = setup();
    const { container } = render(
      <FormField type="radio-group" {...defaultProps} name="Companies" />
    );
    const radioGroup = container.getElementsByClassName("RadioGroup");
    expect(radioGroup).toHaveLength(1);
  });
});
