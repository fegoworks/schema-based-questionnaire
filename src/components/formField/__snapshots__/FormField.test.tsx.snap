// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component - FormField > renders FormField component 1`] = `
<div>
  <div
    class="TextField"
  >
    <input
      id="name"
      name="name"
      role="textbox"
      type="text"
      value=""
    />
  </div>
</div>
`;
